`throw_ref` should accept valid subtypes of `exnref`

danlliu opened the pull request:
#### 169990f7bc448b45efebe8d61409647d12f4892d
<pre>
`throw_ref` should accept valid subtypes of `exnref`
<a href="https://bugs.webkit.org/show_bug.cgi?id=293211">https://bugs.webkit.org/show_bug.cgi?id=293211</a>
<a href="https://rdar.apple.com/151649314">rdar://151649314</a>

Reviewed by Mark Lam.

Currently the WasmFunctionParser will validate a Wasm module by
checking if the operand to `throw_ref` is exactly type `exnref`
(`(ref null exn)`). However, this will not match subtypes which
are equivalent by the Typed Function References spec. We should
be checking if a type such as `(ref exn)` is a valid subtype of
`exnref`, and if so, allowing the module to validate.

The subtyping rules of Typed Function References are at
<a href="https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#reference-types-1">https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#reference-types-1</a>

This patch uses the rule
```
Non-null types are subtypes of possibly-null types

(ref &lt;heaptype1&gt;) &lt;: (ref null &lt;heaptype2&gt;)
iff &lt;heaptype1&gt; &lt;: &lt;heaptype2&gt;
```

* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):

Canonical link: <a href="https://commits.webkit.org/295307@main">https://commits.webkit.org/295307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ea4d0792176f92bf567ba4a2a9db30316d5f876

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109865 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55325 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32909 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79476 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107657 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/19263 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94456 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59783 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19031 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12530 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54697 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97332 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88730 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12583 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112255 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103269 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31816 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23419 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88561 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32180 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90684 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88177 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22476 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33076 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31741 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37091 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31533 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33092 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me.  Please add a link to the relevant section of the Typed Function References spec in your commit message.  Thanks.

danlliu committed:
`throw_ref` should accept valid subtypes of `exnref`
https://bugs.webkit.org/show_bug.cgi?id=293211
rdar://151649314

Reviewed by Mark Lam.

Currently the WasmFunctionParser will validate a Wasm module by
checking if the operand to `throw_ref` is exactly type `exnref`
(`(ref null exn)`). However, this will not match subtypes which
are equivalent by the Typed Function References spec. We should
be checking if a type such as `(ref exn)` is a valid subtype of
`exnref`, and if so, allowing the module to validate.

The subtyping rules of Typed Function References are at
https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#reference-types-1

This patch uses the rule
```
Non-null types are subtypes of possibly-null types

(ref <heaptype1>) <: (ref null <heaptype2>)
iff <heaptype1> <: <heaptype2>
```

* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):

Canonical link: https://commits.webkit.org/295307@main
